include/irr/asset/CTexturePacker.h

```diff
@@ -4,35 +4,96 @@
 #include "irr/asset/format/EFormat.h"
 #include "irr/asset/ICPUImage.h"
 #include "irr/core/alloc/GeneralpurposeAddressAllocator.h"
+#include "irr/core/math/morton.h"
 
 namespace irr {
 namespace asset
 {
 
+//TODO move this to different file
+template<typename addr_type>
+class CSquareAddressAllocator
+{
+protected:
+    using lin_addr_alctr_t = core::GeneralpurposeAddressAllocator<addr_type>;
+
+    const uint32_t ADDR_LAYER_SHIFT;
+    const uint32_t ADDR_MORTON_MASK;
+
+    core::smart_refctd_dynamic_array<uint8_t> m_alctrReservedSpace;
+    lin_addr_alctr_t m_addrAlctr;
+
+public:
+    CSquareAllocator(uint32_t _addrBitCntPerDim, uint32_t _squareSz) : 
+        ADDR_LAYER_SHIFT(_addrBitCntPerDim*2u),
+        ADDR_MORTON_MASK((1u<<ADDR_LAYER_SHIFT)-1u),
+        m_alctrReservedSpace(core::make_refctd_dynamic_array<decltype(m_alctrReservedSpace)>(
+            lin_addr_alctr_t::reserved_size(_squareSz*_squareSz, _squareSz*_squareSz, 1ull))
+        ),
+        m_addrAlctr(reinterpret_cast<void*>(m_alctrReservedSpace->data()), 0u, 0u, _squareSz*_squareSz, _squareSz*_squareSz, 1u)
+    {}
+
+    inline addr_type alloc_addr(size_t _subsquareSz)
+    {
+        const size_t cnt = _subsquareSz*_subsquareSz;
+        return m_addrAlctr.alloc_addr(cnt*cnt, cnt*cnt);
+    }
+
+    inline addr_type unpackAddress_x(addr_type _addr) const
+    {
+        return core::morton2d_decode_x(_addr&ADDR_MORTON_MASK);
+    }
+    inline addr_type unpackAddress_y(addr_type _addr) const
+    {
+        return core::morton2d_decode_y(_addr&ADDR_MORTON_MASK);
+    }
+    inline addr_type unpackAddress_layer(addr_type _addr) const
+    {
+        return (_addr>>ADDR_LAYER_SHIFT);
+    }
+};
+
 class CTexturePacker
 {
+    _IRR_STATIC_INLINE_CONSTEXPR uint32_t TILE_PADDING = 8u;
+
+    _IRR_STATIC_INLINE_CONSTEXPR uint32_t ADDR_LAYER_SHIFT = 28u;//14bit 2d page address and 4bit for 3rd address dimension (layer)
+    _IRR_STATIC_INLINE_CONSTEXPR uint32_t ADDR_MORTON_MASK = (1u<<ADDR_LAYER_SHIFT)-1u;
+
     using addr_alctr_t = core::GeneralpurposeAddressAllocator<uint32_t>;
 
+    static uint32_t megaimgLayerSize(uint32_t _pgSz, uint32_t _tilesPerDim)
+    {
+        return _tilesPerDim*(TILE_PADDING + _pgSz) + TILE_PADDING;
+    }
+    static core::vector2du32_SIMD pageCoords(uint32_t _x, uint32_t _y, uint32_t _pgSz)
+    {
+        core::vector2du32_SIMD coords(_x,_y);
+        coords *= (_pgSz + TILE_PADDING);
+        coords += TILE_PADDING;
+        return coords;
+    }
+
 public:
     struct STexOffset
     {
-        uint32_t x, y;
+        uint16_t x, y;
     };
 
-    CTexturePacker(E_FORMAT _format, uint32_t _szxy = 1u<<14, uint32_t _pgSzxy = 256u) :
+    CTexturePacker(E_FORMAT _format, uint32_t _pgSzxy = 256u, uint32_t _tilesPerDim = 32u, uint32_t _numLayers = 4u) :
```

Comment:
format class not format, generally speaking textures of the same texel block size are compatible with each other for use in ImageViews

Reply:
still upholding my comment about format class

Reply:
still relevant